apps/backend/src/lambdas/report.ts

```diff
@@ -1,16 +1,268 @@
+            cashDepositDatesForLocations.indexOf(event.period.to) - 1
+          ],
+      };
+      console.log(cashDates);
+      if (cashDates.minDate) {
+        const paymentsMatchedOrException =
+          await paymentService.findCashPayments(
+            cashDates,
+            [l.location_id],
+            [MatchStatus.MATCH, MatchStatus.EXCEPTION]
+          );
+        currentCashPayments.push({
+          payments: paymentsMatchedOrException,
+          dateRange: cashDates,
+        });
+      }
+    }
+  }
+  return {
+    cashDeposits: {
+      pending: pendingAndInProgressDeposits,
+      current: currentCashDeposits,
+    },
+    cashPayments: {
+      pending: pendingAndInProgressPayments,
+      current: currentCashPayments,
+    },
+  };
+};
+
+const posReportData = async (
+  app: INestApplicationContext,
+  event: ReportConfig,
+  locations: NormalizedLocation[]
+) => {
+  const paymentService = app.get(PaymentService);
+  const posDepositService = app.get(PosDepositService);
+
+  const pendingAndInProgressPayments = await paymentService.findPosPayments(
+    { minDate: event.period.from, maxDate: event.period.to },
+    locations.map((itm) => itm.location_id),
+    [MatchStatus.PENDING, MatchStatus.IN_PROGRESS]
+  );
+
+  const matchedPaymentsAndExceptions = await paymentService.findPosPayments(
+    { minDate: event.period.to, maxDate: event.period.to },
+    locations.map((itm) => itm.location_id),
+    [MatchStatus.MATCH, MatchStatus.EXCEPTION]
+  );
+
+  const matchedPayments = matchedPaymentsAndExceptions.filter(
+    (itm: PaymentEntity) => itm.status === MatchStatus.MATCH
+  );
+
+  const paymentExceptions = matchedPaymentsAndExceptions.filter(
+    (itm: PaymentEntity) => itm.status === MatchStatus.EXCEPTION
+  );
+
+  const pendingAndInProgressDeposits = await posDepositService.findPosDeposits(
+    { minDate: event.period.from, maxDate: event.period.to },
+    event.program,
+    locations.flatMap((itm: NormalizedLocation) => itm.merchant_ids),
+    [MatchStatus.PENDING, MatchStatus.IN_PROGRESS]
+  );
+  console.log(pendingAndInProgressDeposits, 'PENDING AND IN PROGRESS');
+  const matchedDepositsAndExceptions = await posDepositService.findPosDeposits(
+    { minDate: event.period.to, maxDate: event.period.to },
+    event.program,
+    locations.flatMap((itm: NormalizedLocation) => itm.merchant_ids),
+    [MatchStatus.MATCH, MatchStatus.EXCEPTION]
+  );
+  const matchedDeposits = matchedDepositsAndExceptions.filter(
+    (itm: POSDepositEntity) => itm.status === MatchStatus.MATCH
+  );
+  const depositExceptions = matchedDepositsAndExceptions.filter(
+    (itm: POSDepositEntity) => itm.status === MatchStatus.EXCEPTION
+  );
+  const matchedDepositIdsFromPayments =
+    matchedPayments.map((itm: PaymentEntity) => itm.pos_deposit_match) ?? [];
+  /*eslint-disable */
+  const matchedDepositsFromAnotherDay: POSDepositEntity[] = [];
+
+  matchedDepositIdsFromPayments.forEach((itm: POSDepositEntity | undefined) => {
```

Comment:
You could filter `matchedDepositIdsFromPayments` beforehand to avoid `| undefined`

Reply:
I think since `pos_deposit_match` is an optional field on the PaymentEntity class it will always be possibly undefined? I tried to filter but had the same error. Happy to chat to get some direction on this...